Move fuel and engine parameters to JSON

mlangsdorf opened the pull request:
SUMMARY Infrastructure "Move fuel and engine parameters into JSON"

This refactors much of the vehicle code that deals with the interaction of engines and fuel types.
It shouldn't change anything in the game, but should make it much easier for mods to define their
own ENGINE parts, as most of the remaining logic that was hard-coded on comparisons to a few 
fuel types has been replaced with comparisons to flags and parameters from JSON. These values
are usually attached to the engine vehiclepart, but some are drawn from the fuel.

It's a fairly large commit series, but I've tried to break it up into small, logical commits for easy review
and I could split it into two or more series if that would be helpful.

Fixes #24600

Night-Pryanik commented on src/veh_type.h:
```diff
@@ -78,13 +79,31 @@ enum vpart_bitflags : int {
+
+struct vpslot_engine {
+    float backfire_threshold = 0;
+    int  backfire_freq = 1;
```
The same here.

mlangsdorf replied:
Fixed.

Night-Pryanik commented on src/veh_type.h:
```diff
@@ -206,6 +224,17 @@ class vpart_info
+         *
+         */
+        std::vector<std::string> engine_excludes() const;
+        int  engine_m2c() const;
```
The same here.

mlangsdorf replied:
Fixed.

Night-Pryanik commented on src/vehicle.cpp:
```diff
@@ -3391,20 +3398,22 @@ void vehicle::noise_and_smoke( double load, double time )
-        if( part_flag(p, "MUFFLER") ) {
-            m = 1.0 - (1.0 - part_info(p).bonus / 100.0) * double( parts[p].hp() ) / part_info(p).durability;
+        if( part_flag( p, "MUFFLER" ) ) {
+            m = 1.0 - ( 1.0 - part_info( p ).bonus / 100.0 ) * double( parts[p].hp() ) / part_info(
```
Maybe move the whole `part_info( p ).durability` to the next line?

mlangsdorf replied:
Refactored the entire concept.
Fixed.

mlangsdorf commented on doc/JSON_INFO.md:
```diff
@@ -585,7 +586,15 @@ Mods can modify this via `add:traits` and `remove:traits`.
+"backfire_threshold": 0.5,    // Optional field, defaults to 0. If greater than 0, indicates minimum ratio of damaged to max HP to trigger backfires
+"backfire_freq": 20,          // Optional field unless backfire threshold > 0, then mandatory, defaults to 0. One in X chance of a backfire.
+"noise_factor": 15,           // Optional field, defaults to 0. Multiple engine power by this number to declare noise.
+"damaged_power_factor": 0.5,  // Optional field, defaults to 0. If more than 0, power when damage is scaled to power * ( damaged_power_factor + ( 1 - damaged_power_factor ) * ( damaged HP / max HP )
```
Fixed.

mlangsdorf commented on doc/JSON_INFO.md:
```diff
@@ -585,7 +586,15 @@ Mods can modify this via `add:traits` and `remove:traits`.
+"backfire_freq": 20,          // Optional field unless backfire threshold > 0, then mandatory, defaults to 0. One in X chance of a backfire.
+"noise_factor": 15,           // Optional field, defaults to 0. Multiple engine power by this number to declare noise.
+"damaged_power_factor": 0.5,  // Optional field, defaults to 0. If more than 0, power when damage is scaled to power * ( damaged_power_factor + ( 1 - damaged_power_factor ) * ( damaged HP / max HP )
+"muscle_power_factor": 0,     // Optional field, defaults to 0. If more than 0, each point of the survivor's ST over 8 adds this much power to the engine.
```
Fixed.

mlangsdorf commented on doc/JSON_INFO.md:
```diff
@@ -585,7 +586,15 @@ Mods can modify this via `add:traits` and `remove:traits`.
+"noise_factor": 15,           // Optional field, defaults to 0. Multiple engine power by this number to declare noise.
+"damaged_power_factor": 0.5,  // Optional field, defaults to 0. If more than 0, power when damage is scaled to power * ( damaged_power_factor + ( 1 - damaged_power_factor ) * ( damaged HP / max HP )
+"muscle_power_factor": 0,     // Optional field, defaults to 0. If more than 0, each point of the survivor's ST over 8 adds this much power to the engine.
+"exclusions": [ "souls" ]     // Optional field, defaults to empty. A list of words. A new engine can't be installed on an engine if any engine on the vehicle shares a word from exclusions.
```
Sheesh.
Fixed.

mlangsdorf commented on src/itype.h:
```diff
@@ -383,11 +383,21 @@ struct islot_wheel
-struct islot_fuel
-{
+struct fuel_explosion {
+    int explosion_chance_hot ;
```
Fixed.

mlangsdorf commented on src/veh_type.cpp:
```diff
@@ -645,6 +675,45 @@ int vpart_info::repair_time( const Character &ch ) const
+    return has_flag( VPFLAG_ENGINE ) ? engine_info->backfire_threshold : false;
+}
+
+int  vpart_info::engine_backfire_freq() const
```
Fixed.

OrenAudeles commented on src/vehicle.cpp:
```diff
@@ -874,6 +874,26 @@ bool vehicle::has_engine_type_not( const itype_id &ft, bool const enabled ) cons
+    std::vector<std::string> new_excludes = possible_conflict->engine_excludes();
+    bool has_conflict = false;
+
+    for( size_t e = 0; e < engines.size(); ++e ) {
```
Before iterating through all engines, test if `new_excludes` is empty. If it is empty there is no possibility of conflict and iterating is pointless.

mlangsdorf replied:
Thanks, good catch.

mlangsdorf replied:
Fixed.

OrenAudeles commented on src/vehicle.cpp:
```diff
@@ -3413,57 +3421,55 @@ void vehicle::noise_and_smoke( double load, double time )
+            ( is_engine_type( e, fuel_type_muscle ) || fuel_left( part_info( p ).fuel_type ) ) ) {
             double pwr = 10.0; // Default noise if nothing else found, shouldn't happen
-            double max_pwr = double(power_to_epower(part_power(p, true)))/40000;
+            double max_pwr = double( power_to_epower( part_power( p, true ) ) ) / 40000;
```
Consider splitting off `power_to_epower` into its own variable. Calculation of its parameter `part_power(...)` is complex enough to be worth saving independently. Calculated a second time to assign variable `double j` in combustion branch.

mlangsdorf replied:
Fixed. Gah that code is confusing.

OrenAudeles commented on src/vehicle.cpp:
```diff
@@ -5860,39 +5866,25 @@ int vehicle::break_off( int p, int dmg )
 
-    const fuel_explosion &data = iter->second;
-    const int pow = 120 * (1 - exp(data.explosion_factor / -5000 * (parts[p].ammo_remaining() * data.fuel_size_factor)));
+    const int pow = 120 * ( 1 - exp( data.explosion_factor / -5000 * ( parts[p].ammo_remaining() *
```
`ammo_remaining()` is relatively expensive to calculate (several set inclusion tests), and `exp(...)` is also a bit expensive. Move to inside the `if( one_in(...) )` block's scope as that is the only use of the calculate value `pow` outside of the commented-out debugmsg.

mlangsdorf replied:
that's not really my code, but sure, let's clean it up while I'm here.

mlangsdorf replied:
Thinking about, the compiler should be smart enough to reorder it, and even if doesn't, any out of order processor is automatically going to reorder it. But it's easy enough to change even if I expect it won't have much effect.

mlangsdorf replied:
Also, fixed.

mlangsdorf commented on src/vehicle.cpp:
```diff
@@ -3413,57 +3421,55 @@ void vehicle::noise_and_smoke( double load, double time )
+            } else if( is_engine_type( e, fuel_type_battery ) ) {
+                pwr = cur_pwr * part_info( p ).engine_noise_factor();
+            } else if( is_engine_type( e, fuel_type_muscle ) ) {
+                pwr = cur_pwr * part_info( p ).engine_noise_factor();
```
actually, those test branches were supposed to die in a fire.
<lights match>
okay, they have died in a fire.
Fixed.

mlangsdorf commented:
Squashed, amended, and rebased the commit set, so a couple of changes that were supposed to be in separate commits are separated, and the requested fixes are in the appropriate commits.

If you want to pull the branch for testing and had pulled it already, you'll need to reset your branch to avoid merge errors.

Epictyphlosion commented:
Very interesting. Backfires are loud spurts coming from the exhaust, right? (E.g. anti-lag turbo)

mlangsdorf commented:
Yeah, backfires are noise and smoke.
This PR isn't supposed to change anything. It's a code cleanup that will enable changes like multifuel engines, solid fuel engines, and installing engines as generators in the future.
